Instructor: edit feedback session: load the page in edit mode #1045

HirdayGupta opened the pull request:
Fixes #1045 

**Outline of Solution:**

- In `readyFeedbackEditPage()` (called from `$(document).ready(function())`), removed call to `disableEditFS` and added call to `enableEditFS()`.
- Remove `readonly` property for tinyMCE editor on page load to make Session Instructions editable as well.

**Pending: (Help/Input needed from senior devs):**

- Requesting input on whether the page should load in edit mode even after the page reloads because of question edit (current behavior) or if it should detect the reload and not load in edit mode. If the second option is preferable, requesting assistance in implementing the behavior. (Mainly, I can't seem to identify the block of code that listens for the data sent via the `RedirectResult` in the associated action and adds the status message div to the DOM).

- Also, need help with ironing out the edits to the `InstructorFeedbackEditPageUiTest.java` on two accounts:
    1. Where to check if the page indeed loads in edit mode (`testContent()` or a separate method)
    2. How to prevent the timeout and no such element exceptions occurring due to the new behaviour and the repeated calls to `feedbackEditPage.reloadPage()`.

HirdayGupta commented:
@whipermr5 does the `s.Ongoing` label imply "waiting for help/advice?" If not could you change the label to something that does? 😃 

I ask because I cannot move forward on this without dev input on the questions I've asked above.

whipermr5 commented:
@HirdayGupta Oops sorry I didn't realise! Will take a look at it tomorrow 😅

whipermr5 commented:
> Requesting input on whether the page should load in edit mode even after the page reloads because of question edit (current behavior) or if it should detect the reload and not load in edit mode.

I believe the second option is preferable - the session details should only be editable if the page was accessed via the instructor feedback sessions page. @damithc can confirm?

To implement, I suggest adding a GET parameter (e.g. `editsessiondetails`) to the action that influences the initial behaviour. When the page is accessed via the feedback sessions page, it can be `true`. I.e. the edit button on the feedback sessions page should link to `instructorFeedbackEditPage?editsessiondetails=true&...` whereas when the page is accessed via a redirect, the action that produced the redirect should edit the parameter to false, and the page should redirect to `instructorFeedbackEditPage?editsessiondetails=false&...` (alternatively, the action can drop the parameter altogether, and the default behaviour would be to load in disabled mode).

HirdayGupta commented:
@whipermr5 implemented the GET variable behaviour you recommended. It is set to true only in two cases: when being accessed from feedback sessions page and from the `edit link` in the feedback results page. It is false in all other cases.

Now, I would like to what is the TEAMMATES standard for reading GET variables. Where and how should I be looking for the `editsessiondetails` variable? (I would prefer to do it in JS as that is where I will be controlling the edit mode, but I can't seem to find examples of this)

whipermr5 commented:
@HirdayGupta I believe we add it into the DOM (e.g. as an HTML data attribute or class) in the JSP, then read from it with jQuery.

HirdayGupta commented:
> add it into the DOM (e.g. as an HTML data attribute or class) in the JSP

Really sorry @whipermr5 but could you provide some additional guidance on how I can do this? Or maybe point me to an example in the code?

whipermr5 commented:
@HirdayGupta Use `getRequestParamValue`, like in every other `Action` that reads request parameters.

whipermr5 commented on src/test/resources/pages/instructorFeedbackEditEmpty.html:
```diff
@@ -91,12 +91,12 @@
         <div class="col-sm-12">
           <span class="pull-right">
-            <a class="btn btn-primary btn-sm enable-edit-fs" data-original-title="Edit feedback session details" data-placement="top" data-toggle="tooltip" id="fsEditLink" title="">
+            <a class="btn btn-primary btn-sm enable-edit-fs" data-original-title="Edit feedback session details" data-placement="top" data-toggle="tooltip" id="fsEditLink" style="display: none;" title="">
```
Example of a data attribute in the DOM: `data-original-title="Edit feedback session details"`. You'll need to find the JSP that generates this page and edit it to insert such an attribute (I suggest in the `<form>` element).

HirdayGupta commented:
@whipermr5 I have implemented the required GET variable behaviour. Will now try and sort out the Travis build. Meanwhile, the `InstructorFeedbackEditPageUiTest` is going to have to undergo some structural changes. I will look into this after sorting out all other testing issues and will ask for clarifications soon.

HirdayGupta commented:
@whipermr5 In many of the failing *ActionTests, I am getting `AssertionErrors` because of `AssertEquals()` statements. Is there a way other than manually changing all of these in the tests to fix this part of the tests?

whipermr5 commented:
@HirdayGupta I don't think so!

HirdayGupta commented:
@whipermr5 check out https://github.com/TEAMMATES/teammates/pull/7615/commits/4d8a9b0efde7fa2516bc6a366e8d3d022562121b

That is how I will be solving most of the remaining build errors. If you think that is okay then I shall go ahead with the same approach for the remaining 24.

whipermr5 commented:
@HirdayGupta Would it be possible to use a boolean instead of a string?

HirdayGupta commented:
@whipermr5 Should I also change the paramater to a string in https://github.com/HirdayGupta/teammates/commit/2b6ccdf254a0cba1e2e57d3e7334c8bf39d4b339

I had kept it as string because the other params in `getInstructorFeedbackEditLink()` were strings (which I am now realising was not too smart :p)

whipermr5 commented:
@HirdayGupta Yep!

HirdayGupta commented:
@whipermr5 It's been quite the interesting night for me as you can see from the commit history. :p

So essentially I kept fixing the build errors as Travis kept throwing them at me. (God mode runs, updating URL builders, etc) Requesting you to check up on a70db57 not sure why that happened or what caused it.

I think I am now down to my final few tests that need fixing. One is definitely the InstructorFeedbackEditPageUiTest. Since I figured this is where I would be testing the issue, I decided to hold of on fixing the build for it. Would you have any suggestions on how I could test this feature?

whipermr5 commented:
That test seems to be unstable after #7641 was merged. @Mynk96 do you have any ideas why the ordering sometimes changes?

whipermr5 commented:
> I think I am now down to my final few tests that need fixing. One is definitely the InstructorFeedbackEditPageUiTest. Since I figured this is where I would be testing the issue, I decided to hold of on fixing the build for it. Would you have any suggestions on how I could test this feature?

Do propose something first, then I'll assess it :)

HirdayGupta commented:
@whipermr5 How about this:

On loading the page, verify that it is in edit mode, then click the save button and resume the test as per normal. And then for each action on the page (edit question, add question, copy, delete, etc) test that the FS edit page loads in the expected state.

whipermr5 commented:
> On loading the page, verify that it is in edit mode

Sounds good!

> then click the save button and resume the test as per normal.

Is clicking the save button necessary?

HirdayGupta commented:
@whipermr5 My bad, clicking the save button is not necessary. The test continues without it. Will continue with the rest of my proposed suggestion and upload commits for review. Thanks.

HirdayGupta commented:
@whipermr5 Added an assertion that page is in edit mode on launch. For checking the state of the page after other actions, I think the expected HTML should cover it. Do you think it would be necessary to add explicit `assertTrue(feedbackEditPage.verifyverifyEditSessionBoxIsEnabled()` in addition to the expected HTML tests?

Also, regarding the repositioning of the testEditSessionLink(), I repositioned it because now that the page loads in edit mode, it was throwing a timeout excpetion as it couldn't find the edit button. I placed it below a test case that clicks the `save session` button, so that it could successfully test the link. An alternative to repositioning would be to click the save button right before checking for visibiltiy of edit button in the `testEditSessionLink() ` method. Would you recommend the second approach I sugeested instead of the one I committed in 01566ca?

HirdayGupta commented:
@whipermr5 Found a slight bug 😅 On loading the page with `editsessiondetails=false` the page loads in the correct mode (ie not editable) but the session instructions for some reason were still editable. Found that the cause and removed it in https://github.com/TEAMMATES/teammates/pull/7615/commits/6849b2cb068623a4dd283e8fea595108735c33f5. Then ran god mode to update expected html files. Please review this as well. Thanks! :)

whipermr5 commented on src/main/webapp/dev/js/main/instructorFeedbackEdit.es6:
```diff
@@ -1087,7 +1087,11 @@ function readyFeedbackEditPage() {
 
     // Additional formatting & bindings.
-    disableEditFS();
+    if ($('input[name="editsessiondetails"]').prop('value') === 'true') {
```
Place `editsessiondetails` in `const.es6`.

whipermr5 commented on src/main/webapp/jsp/instructorFeedbackEdit.jsp:
```diff
@@ -36,6 +40,7 @@
   </c:if>
    <br>
   <input type="hidden" id="num-questions" value="${fn:length(data.qnForms)}">
+  <input type="hidden" name="${FEEDBACK_SESSION_ENABLE_EDIT}" value="${data.isToBeLoadedInEditMode}">
```
Is there a reason why you chose to embed the data in an `input` field? This means the parameter is passed around when the form is submitted - is it needed?

HirdayGupta replied:
@whipermr5 No actually, I was just going by the other examples 😅. Can I place it in a div?

whipermr5 replied:
Yup a data attribute can be placed anywhere actually. Maybe place it in the main `div` of `feedbackSessionsForm.tag`?

HirdayGupta replied:
@whipermr5 When I place it in `feedbackSessionsForm.tag`, I get a server error (500) on the `instructorFeedbackSessionsPage`. I get this error because the `feedbackSessionsForm.tag` is reused here, but since there is no `isToBeLoadedInEditMode` parameter being sent to _that_ page (its only sent  to `instructorFeedbackEditPage`), the following div that I inserted gives the `500` error:

```HTML
<div name="${FEEDBACK_SESSION_ENABLE_EDIT}" data-value="${data.isToBeLoadedInEditMode}"></div>
```

Is there some way to avoid this error? Or would it be okay if I kept the current placement (ie: kept it in `instructorFeedbackEdit.jsp`) but changed the implementation from an `<input>` to a `<div></div>`

whipermr5 replied:
Make it an optional parameter, and set the data attribute only if the parameter is defined.

HirdayGupta replied:
@whipermr5 Apologies but how would I do that? Could you point me to some example of an optional parameter?

whipermr5 replied:
Oops, I meant optional tag attribute. Take a look at the top of `feedbackSessionsForm.tag` for an example of a required attribute.

whipermr5 commented on src/main/java/teammates/ui/pagedata/InstructorFeedbackEditPageData.java:
```diff
@@ -42,7 +43,7 @@ public InstructorFeedbackEditPageData(AccountAttributes account, String sessionT
                      Map<String, Boolean> questionHasResponses,
                      List<StudentAttributes> studentList, List<InstructorAttributes> instructorList,
-                     InstructorAttributes instructor) {
+                     InstructorAttributes instructor, String loadInEditMode) {
```
This can be a boolean too.

HirdayGupta replied:
@whipermr5 When I change this, line 80 in `InstructorFeedbackEditPageAction` gives an error because im passing a string. This was the string i received from the GET params, would you recommend parsing the GET param to boolean in the action?

whipermr5 replied:
Yes; read my [comment](https://github.com/TEAMMATES/teammates/pull/7615#discussion_r126856756) below.

whipermr5 commented on src/main/java/teammates/ui/controller/InstructorFeedbackEditPageAction.java:
```diff
@@ -24,6 +24,8 @@ protected ActionResult execute() throws EntityDoesNotExistException {
         String feedbackSessionName = getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
         Assumption.assertPostParamNotNull(Const.ParamsNames.FEEDBACK_SESSION_NAME, feedbackSessionName);
+        String loadInEditMode = getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_ENABLE_EDIT);
+        Assumption.assertPostParamNotNull(Const.ParamsNames.FEEDBACK_SESSION_ENABLE_EDIT, loadInEditMode);
```
I feel that we should allow it to be `null` so that the `editsessiondetails` part of the URL parameters becomes optional (as it is now we have way too many parameters littering the URL - the usual practice for web apps should be to hide non-configurable parameters from the user such that they see a clean URL). If `"true".equals(loadInEditMode)`, set `loadInEditMode` to `true`, else `false` (whether it's `null` or other user-input value like `0`, `false` etc.).

whipermr5 commented:
By the way, the suspicious change seen in fb4f61c is a manifestation of #7669.

whipermr5 commented:
As for the removal of `initEditor` in 6849b2c, when will `initEditor` be called? It seems to me that instead of removing it there, the other places where `initEditor` should be reviewed. The `readOnly` flag of `initEditor` should depend on `editsessiondetails`.

HirdayGupta commented:
@whipermr5 Here's the thing, the `initEditor` will now definitely be called either in `disableEditFS()` or `enableEditFS()` (depending on `editsessiondetails`) in `readyFeedbackEditPage()` in the `document.ready()` function. 

Before I removed the additional call to `initEditor` before `readyFeedbackEditPage()` in 6849b2c, it was being called twice each time on page load. And when `editsessiondetails=false`, for some reason, the `readonly` parameter would not be overwritten to false by the call to `disableEditFS()`. So, removing this first, unnecessary, call ensures that it is called only once for the `#instructions` field.

whipermr5 commented:
@HirdayGupta Oh I see; it's fine then. Didn't realise 😅

whipermr5 commented on src/main/java/teammates/ui/controller/InstructorFeedbackEditPageAction.java:
```diff
@@ -25,8 +25,11 @@ protected ActionResult execute() throws EntityDoesNotExistException {
-        Assumption.assertPostParamNotNull(Const.ParamsNames.FEEDBACK_SESSION_ENABLE_EDIT, loadInEditMode);
+        String loadInEditModeParam = getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_ENABLE_EDIT);
+        boolean loadInEditMode = false;
+        if (loadInEditModeParam != null && "true".equals(loadInEditModeParam)) {
```
Writing `"true".equals(loadInEditModeParam)` already takes care of the null case. The whole thing can be simplified to
```
boolean loadInEditMode = "true".equals(loadInEditModeParam);
```

HirdayGupta replied:
Ahh, I see. My bad, I thought that approach would throw a `NullPointerException`. Didn't realise that using the string that may be null as the param (instead of before the dot operator) would handle this.

HirdayGupta commented:
@whipermr5 All done (that took a while, I'm sorry about that). Ready for review!

HirdayGupta commented:
@whipermr5 Can you restart the build? Something seems to have gone wrong with downloading appengine resources.

whipermr5 commented:
@HirdayGupta Pending some more HTML test file updates :)

HirdayGupta commented:
@whipermr5 I had noticed haha. My laptop was working on it. :p (It takes a surprisingly long time to run the `./gradlew appengineRun CiTests` locally!)

All done now. Build should hopefully pass. Also, requesting you to read the description on e28f256. Was manually editing the test HTML file manually the right thing to do, given that I was aware of the unwanted change from https://github.com/TEAMMATES/teammates/commit/fb4f61c85ac71ba8968ca3ff65f1cc488b108c24 and that the linked issue (#7669) was still open?

whipermr5 commented:
> Also, regarding the repositioning of the testEditSessionLink(), I repositioned it because now that the page loads in edit mode, it was throwing a timeout excpetion as it couldn't find the edit button. I placed it below a test case that clicks the save session button, so that it could successfully test the link.

Oops realised I forgot to reply to this. It's fine 👍

> Was manually editing the test HTML file manually the right thing to do, given that I was aware of the unwanted change from fb4f61c and that the linked issue (#7669) was still open?

Yup that was the right thing to do!

whipermr5 commented on src/main/java/teammates/common/util/AppUrl.java:
```diff
@@ -42,4 +42,8 @@ public AppUrl withQuestionNumber(String questionNumber) {
         return withParam(Const.ParamsNames.FEEDBACK_QUESTION_NUMBER, questionNumber);
     }
 
+    public AppUrl withEnableSessionEditDetails(String isLoadedInEditMode) {
```
This can be a boolean too. Convert it to a string before passing to `withParam`.

whipermr5 commented on src/main/java/teammates/ui/controller/InstructorFeedbackEditPageAction.java:
```diff
@@ -25,6 +25,8 @@ protected ActionResult execute() throws EntityDoesNotExistException {
         String feedbackSessionName = getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME);
         Assumption.assertPostParamNotNull(Const.ParamsNames.FEEDBACK_SESSION_NAME, feedbackSessionName);
+        String loadInEditModeParam = getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_ENABLE_EDIT);
+        boolean loadInEditMode = "true".equals(loadInEditModeParam);
```
`shouldLoadInEditMode` to follow our boolean variable naming convention.

whipermr5 commented on src/main/java/teammates/ui/pagedata/InstructorFeedbackEditPageData.java:
```diff
@@ -33,6 +33,7 @@
     private FeedbackQuestionEditForm newQnForm;
     private FeedbackSessionPreviewForm previewForm;
     private String statusForAjax;
+    private String isToBeLoadedInEditMode;
```
I think this can also be a boolean? Also, rename to `shouldLoadInEditMode` as it's more concise :)

whipermr5 commented on src/main/java/teammates/ui/pagedata/InstructorFeedbackEditPageData.java:
```diff
@@ -42,7 +43,7 @@ public InstructorFeedbackEditPageData(AccountAttributes account, String sessionT
                      Map<String, Boolean> questionHasResponses,
                      List<StudentAttributes> studentList, List<InstructorAttributes> instructorList,
-                     InstructorAttributes instructor) {
+                     InstructorAttributes instructor, boolean loadInEditMode) {
```
Rename here too.

whipermr5 commented on src/main/java/teammates/ui/pagedata/InstructorFeedbackEditPageData.java:
```diff
@@ -362,4 +365,8 @@ public boolean getHasError() {
         this.hasError = value;
     }
+
+    public String getIsToBeLoadedInEditMode() {
```
I think this can be a boolean too. It should only finally become a string when the JSP echoes it out in HTML.

whipermr5 commented on src/main/webapp/WEB-INF/tags/instructor/feedbacks/feedbackSessionsForm.tag:
```diff
@@ -278,4 +279,5 @@
             value="${data.account.googleId}">
         <input type="hidden" name="<%= Const.ParamsNames.SESSION_TOKEN%>" value="${data.sessionToken}">
     </form>
+    <div name="<%= Const.ParamsNames.FEEDBACK_SESSION_ENABLE_EDIT %>" data-value="${fsEnableEdit}"></div>
```
Apologies I wasn't clear; the idea I had in mind was to embed the data in a data attribute of an existing element. The `form` element at the very top seems ideal for this purpose. The name of the data attribute can be `data-editsessiondetails`; the JavaScript can then do a `$("#form_feedbacksession").data("editsessiondetails")` (substitute with the relevant constants).

whipermr5 commented on src/test/java/teammates/test/cases/action/InstructorFeedbackAddActionTest.java:
```diff
@@ -219,11 +223,13 @@ protected InstructorFeedbackAddAction getAction(String... params) {
-    protected String getPageResultDestination(
-            String parentUri, String courseId, String fsname, String userId, boolean isError) {
+    protected String getPageResultDestination(String parentUri, String courseId, String fsname,
+            boolean isToBeLoadedInEditMode, String userId, boolean isError) {
```
The good thing about allowing the param to be null is that all these changes are unnecessary, as a null param is interpreted as false. All the changes to this file can be reverted :)

HirdayGupta replied:
@whipermr5 To make this and similar changes, I would need to overload the method being called in [this line](https://github.com/TEAMMATES/teammates/pull/7615/files#diff-962b47c7f33822774b4debc9ddca5a16R82) (Ie `getInstructorFeedbackLink`) to also accept a set of params that does not include the `isToBeLoadedInEditMode` boolean. Would overloading be the best approach here? Or should I use the `boolean... shouldLoadInEditMode` approach as an optional param and then handle the null case using `if (shouldLoadInEditMode[0] == null`) ?

whipermr5 replied:
No need to modify `getInstructorFeedbackLink` in the action - this is only about the `getPageResultDestination` helper method in tests.

HirdayGupta replied:
@whipermr5 I will need to modify the `getIntructorFeedbackLink` because the test fails unless the URL generated from the test helper method and the action destination are the same. See [this line](https://github.com/TEAMMATES/teammates/pull/7615/files/e28f25633cd3a7663cc6cca9c83d7f5a25192b84#diff-64b656ee077cfb7255c4a643b1d83b5fR57).

whipermr5 replied:
Oh I see; my bad 😅 Yes; I think overloading would be good.

whipermr5 commented on src/test/java/teammates/test/cases/action/InstructorFeedbackCopyActionTest.java:
```diff
@@ -199,11 +201,13 @@ protected InstructorFeedbackCopyAction getAction(String... params) {
-    protected String getPageResultDestination(
-            String parentUri, String courseId, String fsname, String userId, boolean isError) {
+    protected String getPageResultDestination(String parentUri, String courseId, String fsname,
+            boolean isToBeLoadedInEditMode, String userId, boolean isError) {
```
Here too.

whipermr5 commented on src/test/java/teammates/test/cases/action/InstructorFeedbackQuestionAddActionTest.java:
```diff
@@ -1001,11 +1020,13 @@ protected InstructorFeedbackQuestionAddAction getAction(String... params) {
-    protected String getPageResultDestination(
-            String parentUri, String courseId, String fsname, String userId, boolean isError) {
+    protected String getPageResultDestination(String parentUri, String courseId, String fsname,
+            boolean isToBeLoadedInEditMode, String userId, boolean isError) {
```
Here too.

whipermr5 commented on src/test/java/teammates/test/cases/action/InstructorFeedbackQuestionCopyActionTest.java:
```diff
@@ -183,11 +186,13 @@ protected InstructorFeedbackQuestionCopyAction getAction(String... params) {
-    protected String getPageResultDestination(
-            String parentUri, String courseId, String fsname, String userId, boolean isError) {
+    protected String getPageResultDestination(String parentUri, String courseId, String fsname,
+            boolean isToBeLoadedInEditMode, String userId, boolean isError) {
```
Here too.

whipermr5 commented on src/test/java/teammates/test/cases/action/InstructorFeedbackQuestionEditActionTest.java:
```diff
@@ -1542,11 +1568,13 @@ protected InstructorFeedbackQuestionEditAction getAction(String... params) {
-    protected String getPageResultDestination(
-            String parentUri, String courseId, String fsname, String userId, boolean isError) {
+    protected String getPageResultDestination(String parentUri, String courseId, String fsname,
+            boolean isToBeLoadedInEditMode, String userId, boolean isError) {
```
Here too.

whipermr5 commented on src/test/resources/pages/instructorFeedbackAllSessionTypes.html:
```diff
@@ -758,6 +758,8 @@ <h5 class="col-xs-2" data-original-title="Select this option to use a custom tim
       <input name="user" type="hidden" value="CFeedbackUiT.instructor">
       <input name="token" type="hidden" value="${sessionToken}">
     </form>
+    <div data-value="" name="editsessiondetails">
```
How come this is empty?

HirdayGupta replied:
@whipermr5 this is empty because we pass an optional attribute down to the `feedbackSessionsForm` tag while loading the instructorFeedbackEditPage. Wherever the value is empty, the page being loaded is the `instructorFeedbackSessionsPage`, which does not get a value for the optional attribute. Hence it is empty. (See https://github.com/TEAMMATES/teammates/pull/7615/commits/55e1f3f73c9a237dea8790c758f96248342ac222 for the optional attribute being added)

whipermr5 replied:
Perhaps make it default to `false` instead?

whipermr5 replied:
Can also make it required if it's only used in one other place.

HirdayGupta replied:
@whipermr5 Since the only other place it is used is the `instructorFeedbackSessionsPage`, where it is in edit mode by default (and indeed should be), I think passing `false` would be misleading. I also don't think it is necessary to pass the variable down there at all. Should I instead put a test and add in the data attribute only if the variable exists?

whipermr5 replied:
Oops I got it mixed up; I meant we should pass `true`.

HirdayGupta replied:
@whipermr5 Is it really necessary to even pass it? The JS for that page has no dependency on it whatsoever.

whipermr5 replied:
Not necessary; I wasn't aware of that 😅 https://github.com/TEAMMATES/teammates/pull/7615#discussion_r129240543

HirdayGupta commented:
@whipermr5 My github desktop spazzed and did some really strange things to my branch. I hit `Update from TEAMMATES/master` and it pointed out sorts of conflicts that weren't actually there. And then when I committed this, it showed up as a regular commit instead of a merge commit. (The commit I am referring to here is 1856c4b). The compile error shown by travis alongside 811dd5d indicates that some very unrelated changes had taken place in 1856c4b. I have fixed the compile error in 61fb127. but I worry that some other minor details may have been changed.

Is there a way for me to revert the last three commits and somehow still merge in the changes from master? I don't know what else 1856c4b changed and there's too many changes in this PR to keep effective track. Any help would

whipermr5 commented:
@HirdayGupta Yup; use reset and force push :)

HirdayGupta commented:
> Also I highly recommend you switch to SourceTree.

@whipermr5 thanks for the suggestion! Shall look into SourceTree and start using it soon (perhaps after the evaluation because then I will have time to play around with new software :p).

> Yup; use reset and force push :)

Just to verify the commands I will be running so that nothing dangerous happens (this PR was a LOT of work!)

```Git
git reset e28f256
git push -f
```

Do the above git commands look okay to you?

HirdayGupta commented:
@whipermr5 These are the steps I performed now (leaving GitHub Desktop completely out of the equation)

1. `git reset --hard e28f256`
2. `git push -f`
3. `git merge --no-ff TEAMMATES/master`
4. There was only one conflict pointed out at the feedbackSessionsForm.tag (view output from git merge [here](https://gist.github.com/HirdayGupta/1158a368d3f17df515097b855992a5a6))
5. I fixed this conflict in my text editor and then:
7. `git add`
8. `git commit -m "Merge remote-tracking branch 'TEAMMATES/master' into 1045-load-edit-fs-in-edit-mode"`
9. `git push`
10. Got the same build error as earlier on Travis. I realised that some new test methods were added which called a method whose signature my PR had changed. This was causing the build errors.
11. Fixed these in https://github.com/TEAMMATES/teammates/pull/7615/commits/da7befd398909d0677ce7284e1e40f31cf2d8172
12. Hopefully now my branch is up-to-date and hasn't lost any of the changes from earlier.

whipermr5 commented on src/main/java/teammates/common/util/AppUrl.java:
```diff
@@ -42,4 +42,8 @@ public AppUrl withQuestionNumber(String questionNumber) {
         return withParam(Const.ParamsNames.FEEDBACK_QUESTION_NUMBER, questionNumber);
     }
 
+    public AppUrl withEnableSessionEditDetails(boolean isLoadedInEditMode) {
```
Rename the boolean here too.

whipermr5 commented on src/main/java/teammates/ui/pagedata/PageData.java:
```diff
@@ -427,10 +427,11 @@ public String getInstructorFeedbackDeleteLink(String courseId, String feedbackSe
     }
 
-    public String getInstructorFeedbackEditLink(String courseId, String feedbackSessionName) {
+    public String getInstructorFeedbackEditLink(String courseId, String feedbackSessionName, boolean loadInEditMode) {
```
Here too.

whipermr5 commented on src/main/webapp/WEB-INF/tags/instructor/feedbacks/feedbackSessionsForm.tag:
```diff
@@ -6,11 +6,13 @@
       action="${fsForm.formSubmitAction}"
-      id="form_feedbacksession">
+      id="form_feedbacksession"
+      data-<%= Const.ParamsNames.FEEDBACK_SESSION_ENABLE_EDIT %> = "${fsEnableEdit}">
```
Delete the spaces surrounding `=`.

whipermr5 commented on src/main/java/teammates/ui/pagedata/InstructorFeedbackEditPageData.java:
```diff
@@ -59,6 +60,8 @@ public void init(FeedbackSessionAttributes feedbackSession, List<FeedbackQuestio
         buildPreviewForm(feedbackSession, studentList, instructorList);
 
+        this.shouldLoadInEditMode = shouldLoadInEditMode;
+
```
This line can be removed.

whipermr5 replied:
Sorry I was unclear; I meant the blank line as it's unrelated. The assignment looks like it's necessary?

HirdayGupta replied:
@whipermr5 yes, the assignment was necessary. My bad, I thought got confused because of the little JSP trick with the getter methods. Have put back the changes. (Note, I had run god mode with these changes. Have also reverted that commit).

whipermr5 commented on src/main/webapp/jsp/instructorFeedbackEdit.jsp:
```diff
@@ -21,9 +21,10 @@
 <c:set var="EMPTY_FEEDBACK_SESSION_MESSAGE">
   <%= Const.StatusMessages.FEEDBACK_QUESTION_EMPTY %>
 </c:set>
+
```
Unrelated change.

HirdayGupta commented:
@whipermr5 Dropping a note that I've cancelled the Travis build as I was pending a god mode run.

whipermr5 commented on src/main/webapp/dev/js/main/instructorFeedbackEdit.es6:
```diff
@@ -1087,7 +1087,11 @@ function readyFeedbackEditPage() {
 
     // Additional formatting & bindings.
-    disableEditFS();
+    if ($('#form_feedbacksession').data(`${ParamsNames.FEEDBACK_SESSION_ENABLE_EDIT}`) === true) {
```
Wouldn't this condition fail if the value of the data attribute is empty?

whipermr5 replied:
I know I originally suggested an optional JSP tag attribute, but I now feel it'd be better to make it required and always explicitly show it in our code. The thing that's optional should be the GET param, as that's displayed to the user right in the URL bar and can be easily modified.

HirdayGupta replied:
@whipermr5 Yes, and it would then load the page with edit mode disabled. That being said, there is no way for this to be empty in the current scenario. This `instructorFeedbackEdit.es6` is loaded only for the `instructorFeedbackEditPage`.

The property will be empty only on the `instructorFeedbackSessionsPage`, whose JS file has no dependence on the attribute.

whipermr5 replied:
Oh I see! That changes things. Maybe use `c:if` to remove the entire data attribute if it's not set then.

whipermr5 commented on src/main/webapp/WEB-INF/tags/instructor/feedbacks/feedbackSessionsForm.tag:
```diff
@@ -6,11 +6,13 @@
       action="${fsForm.formSubmitAction}"
-      id="form_feedbacksession">
+      id="form_feedbacksession"
+      data-<%= Const.ParamsNames.FEEDBACK_SESSION_ENABLE_EDIT %>="${fsEnableEdit}">
```
Just marking the relevant line I was referring to above.

HirdayGupta replied:
> Maybe use c:if to only include the data attribute if the tag attribute is not set then.

Got it. Thats what I was suggesting [here](https://github.com/TEAMMATES/teammates/pull/7615#discussion_r129230073) haha.

Quick question: How do I check the if the property is set?

whipermr5 replied:
https://github.com/TEAMMATES/teammates/search?utf8=%E2%9C%93&q=c%3Aif&type=

HirdayGupta replied:
@whipermr5 Sorry. Should have done that myself. 😅

HirdayGupta commented:
@whipermr5 Ready for review :)

whipermr5 commented on src/test/resources/pages/instructorFeedbackAllSessionTypes.html:
```diff
@@ -6,7 +6,7 @@
   <br>
   <div class="well well-plain">
-    <form action="/page/instructorFeedbackAdd" class="form-group" id="form_feedbacksession" method="post">
+    <form action="/page/instructorFeedbackAdd" class="form-group" data-editsessiondetails="" id="form_feedbacksession" method="post">
```
Do the latest changes correctly skip the attribute here?

HirdayGupta replied:
Looking at the Travis error logs, yes, they do. God mode is just finishing up on my computer. Will commit the updated HTML files soon.
